### PR TITLE
Fixed tier price issue on configurable product

### LINKED ISSUE #38390: Tier Price is not being deleted from Configurable Product
### Preconditions (*)

1. Magento 2.4.6-p2
2. Magento 2.4-develop

### Description (*)
The tier price for the configurable product isn’t deleted when a simple product is converted into a configurable product. The Actual price of a product is removed from the database but the tier price  remains there in `catalog_product_entity_tier_price`. It will show the minimum price for the product at the front end even if any child variant of the product isn’t having any tier price. This can lead to a mismatch in the price. See the attached screenshots below.

Screenshot 1:
Here you can see in the screenshot that the simple product has the tier price in the table `catalog_product_entity_tier_price`.
image

Screenshot 2:
After Converting the simple product into the configurable product by creating the variants, you can see in the below screenshot that the tier price remains there. Even after removing the tier price from all its variants, it remains there in the table `catalog_product_entity_tier_price`.

These things lead in a way to mismatch of the price on a product page.
If you check the product on a category-page/listing-page, you will find the minimum price coming from the tier price available for a  product. But when you visit the product page for the same and check all the variants, you will come to know that there is no minimum price available in any of the child variants of the product.
image
image

### Steps to reproduce (*)

1. Create a simple product with a tier price.
2. Edit that simple product and convert it into a configurable product by creating some variants with a price greater than the tier price and make sure that any variant should not have any tier price.
3. Observe in the database that table `catalog_product_entity_tier_price` still has the entry for the particular product with the tier price.
4. Visit the particular product on a listing page and product page and observe the price.

### Expected result (*)

The tier price of the configurable product should also get deleted when a product is converted from a simple product to a configurable product. It should show the minimum price of the variants instead of a tier price of the parent configurable product at the frontend.
### Actual result (*)

The tier price of the configurable product is not deleted when a product is converted from a simple product to a configurable product. It is showing the tier price of a configurable product as a minimum price.

### PR DESCRIPTION
**Pull request by zeel1012**
### Description (*)

The tier price for the configurable product isn’t deleted when a simple product is converted into a configurable product. The Actual price of a product is removed from the database but the tier price  remains there in `catalog_product_entity_tier_price`. 
### Fixed Issues (if relevant)

1. Fixes  #38390 

### Manual testing scenarios (*)

1. Create a simple product with a tier price.
2. Edit that simple product and convert it into a configurable product by creating some variants with the price greater than the tier price. Variants should not have any tier price.
3. Observe in the database that table `catalog_product_entity_tier_price` still has the entry for the particular product with the tier price.
4. Visit the particular product on a listing page and product page and observe the price.

### Additional Information

File:magento/module-catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandler.php
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandler.php#L83-L92
The problem lies in the fact that the tier price handler is called every time we update the product, but it fails to remove the tier price value from the database when we convert a simple product into a configurable product.

### Solution
To resolve this issue, we need to add a condition. For deleting the tier price from the database, set the $priceRows variable as an empty array. Afterward, in this file, this triggers the condition on line  92 to become true, and a delete action is performed. Before executing this action, it's crucial to check if the product type is configurable. Also, consider whether the simple product has a tier price. After that, a condition is added: If the product type is configurable and the simple product has a tier price, the $priceRows variable is set to an empty array. Following this adjustment, the tier price of the configurable product is removed from the database table.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by zeel1012** [comment]
@magento run all tests

**Comment by zeel1012** [comment]
@magento run all tests

**Review by andrewbess — CHANGES_REQUESTED** [review]
Hello @zeel1012 
Thank you for your contribution
Could you please check and fix test failures
Thank you in advance.

**Comment by zeel1012** [comment]
@magento run all tests

**Comment by zeel1012** [comment]
@magento run all tests

**Comment by zeel1012** [comment]
@magento run all tests

**Comment by zeel1012** [comment]
@andrewbess, Hello, I have updated my code and resolved the issue by using a plugin. Previously, my suggestion faced challenges with static and unit tests due to dependency issues. With this solution using the plugin, the tests now pass successfully. Additionally, I have checked some tests that still fail, but they may not be relevant to my solution.

**Comment by zeel1012** [comment]
@magento run Integration Tests,Functional Tests CE,Functional Tests B2B,Functional Tests EE,WebAPI Tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Unit Tests ,Integration Tests,Functional Tests EE,Functional Tests CE ,Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run Unit Tests ,Functional Tests EE ,Functional Tests CE,Functional Tests B2B,Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Before execute plugin.
+     *
+     * @param UpdateHandler $subject
+     * @param mixed $entity
+     * @param array $arguments
+     * @return array
+     */
+    public function beforeExecute(UpdateHandler $subject, $entity, $arguments = [])
```
Comment (engcom-Hotel) [inline app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php]: Please add the return type as array
```suggestion
    public function beforeExecute(UpdateHandler $subject, $entity, $arguments = []): array
```

File: app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    private $attributeRepository;
+
+    /**
+     * UpdateHandlerPlugin constructor.
+     *
+     * @param ProductAttributeRepositoryInterface $attributeRepository
+     */
+    public function __construct(ProductAttributeRepositoryInterface $attributeRepository)
+    {
+        $this->attributeRepository = $attributeRepository;
```
Comment (engcom-Hotel) [inline app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php:32]: Use [readonly property](https://www.php.net/manual/en/language.oop5.properties.php#language.oop5.properties.readonly-properties)

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @zeel1012,
Thanks for the contribution!

Please add some automated test for this fix in accordance to [DOD](https://developer.adobe.com/commerce/contributor/guides/code-contributions/definition-of-done/#definition-of-done-1).

Also please go through the review points made below.

Thanks

File: app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php (hunk 1 above)
Comment (zeel1012) [reply app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php]: @engcom-Hotel , I have made changes as per your suggestions.

File: app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php (hunk 2 above)
Comment (zeel1012) [reply app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php:32]: @engcom-Hotel , I have made changes as per your suggestions.

**Comment by zeel1012** [comment]
Hello @engcom-Hotel,

I have made changes as per your suggestions. Automated tests still remain.

**Comment by zeel1012** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * Plugin for handling tier prices during product attribute backend update.
+ */
+class UpdateHandlerPlugin
+{
+    /**
+     * @var ProductAttributeRepositoryInterface
+     */
+    public readonly ProductAttributeRepositoryInterface $attributeRepository;
```
Comment (engcom-Hotel) [inline app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php]: Please make this a private
```suggestion
    private readonly ProductAttributeRepositoryInterface $attributeRepository;
```

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Thanks @zeel1012,

We are waiting for the automated tests for this PR. Also please have a look into the below review comment.

Thanks

File: app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php (hunk 3 above)
Comment (zeel1012) [reply app/code/Magento/ConfigurableProduct/Plugin/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandlerPlugin.php]: Hello @engcom-Hotel ,
I have made changes as per your suggestions.

**Comment by zeel1012** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
As the automated is still not implemented hence moving it to the `Changes Requested` bucket.

**Comment by zeel1012** [comment]
Hi @engcom-Hotel ,
I've added an automated unit test as per your suggestions.

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by zeel1012** [comment]
@magento run WebAPI Tests,Integration Tests,Functional Tests EE,Functional Tests CE ,Functional Tests B2B

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Delta** [comment]
### :heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento **`2.4-develop`** 

**Manual testing scenario:**

1.Create a simple product with a tier price.
2.Edit that simple product and convert it into a configurable product by creating some variants with the price greater than 
   the tier price. Variants should not have any tier price.
3.Observe in the database that table catalog_product_entity_tier_price still has the entry for the particular product with the t 
   tier price.
4.Visit the particular product on a listing page and product page and observe the price.

**Before:** :heavy_multiplication_x: 

Screenshot 2024-04-01 at 6 47 56 PM

Screenshot 2024-04-01 at 6 50 24 PM

Screenshot 2024-04-01 at 6 51 56 PM

Screenshot 2024-04-01 at 7 13 28 PM

Screenshot 2024-04-01 at 7 15 05 PM

Screenshot 2024-04-01 at 7 22 22 PM

Screenshot 2024-04-01 at 7 22 49 PM

Screenshot 2024-04-01 at 7 22 07 PM


**After:** :heavy_check_mark:  

Screenshot 2024-04-02 at 1 55 34 PM

Screenshot 2024-04-02 at 1 55 52 PM

Screenshot 2024-04-02 at 1 56 25 PM

Screenshot 2024-04-02 at 1 57 06 PM

Screenshot 2024-04-02 at 1 59 43 PM

Screenshot 2024-04-02 at 2 00 23 PM

Screenshot 2024-04-02 at 2 00 55 PM


Builds are failed. Hence, moving this PR to **Extended Testing.**

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests EE, WebAPI Tests

**Comment by engcom-Charlie** [comment]
The functional B2B failures are different in recent 2 builds with same code. 

**Run 1:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38427/519a0b36504f1bc234b676db45f15109/Functional/allure-report-b2b/index.html
image

**Run 2:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38427/452e552b9217ccf6a5775356442deba2/Functional/allure-report-b2b/index.html#categories

There are open issues for current build's test failures. [ACQE-6363](https://jira.corp.adobe.com/browse/ACQE-6363), [ACQE-6364](https://jira.corp.adobe.com/browse/ACQE-6364)
image

The functional EE failures are different in recent 2 builds with same code. 

**Run 1:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38427/f1c61793490c8b08cea8f66cbbd4d4b8/Functional/allure-report-ee/index.html#categories
image


**Run 2:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38427/4bd896de4c92400f688879d51db6a874/Functional/allure-report-ee/index.html#categories/cbe8f4d320cf77630fd09e78ce65778b/f180181d6e8042ef/
image
There is open issue exits for StorefrontCreateOrderAllQuantityGroupedProductOptionDefaultStockTest fram latest build.

The Web API failures are different in recent 2 builds with same code. There are open issues present for the same. [ACQP-4710](https://jira.corp.adobe.com/browse/ACQP-4710)

**Run 1:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38427/87cad92e7955ff61581ca552c232619c/WebApi/console-error-logs.html
image

**Run 2:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38427/ba6237b637dcca943f7d7eda6bf63c9e/WebApi/console-error-logs.html
image
image

As all failed tests are either flaky or having open issues attached to them, moving this PR to Merge In Progress.

Thank you!

**Pull request merged**